The feed API challenge for Iaroslav Pasternak

Yaro812 opened the pull request.

caiozullo commented:
Hi @Yaro812, please rename the PR following the instructions in the readme file. Thank you!

Yaro812 commented:
Done, thank you!

—

On 12 Sep 2021, 11:20 +0300, Caio Zullo ***@***.***>, wrote:
> Hi @Yaro812, please rename the PR following the instructions in the readme file. Thank you!
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

thiagopenna commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -19,6 +19,30 @@ public final class RemoteFeedLoader: FeedLoader {
+					return completion(.failure(Error.invalidData))
+				}
+
+				do {
+					let feedImages = try self.feedImages(from: data)
+					completion(.success(feedImages))
+				} catch {
+					completion(.failure(Error.invalidData))
+				}
```
Following from my previous comment, by having the `map` function `throw` the correct error we don't have to overwrite it here. That way we could use a `do/catch` statement and complete with the `error` caught.
```swift
	do {
		let feedImages = try self.feedImages(from: data)
		completion(.success(feedImages))
	} catch {
		completion(.failure(error))
	}
```

Or we could use `Result`'s [initializer](https://developer.apple.com/documentation/swift/result/3139399-init) that takes a throwing closure as parameter and returns `.success` with the value returned, or `.failure` with any thrown error:
```suggestion
				completion(Result { try self.feedImages(from: data) })
```

thiagopenna commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -19,6 +19,30 @@ public final class RemoteFeedLoader: FeedLoader {
+		}
+	}
+
+	private func feedImages(from data: Data) throws -> [FeedImage] {
```
We could declare this helper function as `static`, so we won't need a reference to `self` from inside the `load` function.

thiagopenna commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -19,6 +19,30 @@ public final class RemoteFeedLoader: FeedLoader {
 	public func load(completion: @escaping (FeedLoader.Result) -> Void) {
-		fatalError("Must be implemented")
+		client.get(from: url) { [weak self] result in
+			guard let self = self else { return }
```
Following from my previous comment, if we remove the need to upgrade self from a weak reference, we can just check for `nil` instead:
```suggestion
			guard self != nil else { return }
```

thiagopenna commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -19,6 +19,30 @@ public final class RemoteFeedLoader: FeedLoader {
+
+			switch result {
+			case let .success((data, response)):
+				guard response.statusCode == 200 else {
+					return completion(.failure(Error.invalidData))
+				}
```
Consider moving this logic to a helper to keep the `load` function shorter and easier to read. Since this portion of code deals with `mapping` the result of `loading`, consider moving this logic to the already existing helper.

thiagopenna reviewed (commented):
Hi @Yaro812, you almost passed the challenge!

This is Penna, I am an instructor at Essential Developer.

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

Yaro812 commented:
Hi Penna,

Thank you for your review. I addressed all the comments. Please consider reviewing one more time

—

Thank you!
Iaroslav Pasternak
On 16 Sep 2021, 15:14 +0300, thiagopenna ***@***.***>, wrote:
> @thiagopenna commented on this pull request.
> Hi @Yaro812, you almost passed the challenge!
> This is Penna, I am an instructor at Essential Developer.
> Please, check my review comments and let me know once you're done so I can review it again.
> Thank you 🙂
> In FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
> > +
>
> +                               do {
>
> +                                       let feedImages = try self.feedImages(from: data)
>
> +                                       completion(.success(feedImages))
>
> +                               } catch {
>
> +                                       completion(.failure(Error.invalidData))
>
> +                               }
>
> +
>
> +                       case .failure:
>
> +                               completion(.failure(Error.connectivity))
>
> +                       }
>
> +               }
>
> +       }
>
> +
>
> +       private func feedImages(from data: Data) throws -> [FeedImage] {
>
> +               let feedResult = try JSONDecoder().decode(FeedResult.self, from: data)
>
> This function could throw the expected error type instead of JSONDecoder errors. That way, clients of this function can use the thrown error directly.
> ⬇️ Suggested change
> -              let feedResult = try JSONDecoder().decode(FeedResult.self, from: data)
>
> +               guard let feedResult = try? JSONDecoder().decode(FeedResult.self, from: data) else {
>
> +                       throw Error.invalidData
>
> +               }
>
> In FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
> > +                          do {
>
> +                                       let feedImages = try self.feedImages(from: data)
>
> +                                       completion(.success(feedImages))
>
> +                               } catch {
>
> +                                       completion(.failure(Error.invalidData))
>
> +                               }
>
> Following from my previous comment, by having the map function throw the correct error we don't have to overwrite it here. That way we could use a do/catch statement and complete with the error caught.
>     do {
>
>                let feedImages = try self.feedImages(from: data)
>
>                completion(.success(feedImages))
>
>        } catch {
>
>                completion(.failure(error))
>
>        }
> Or we could use Result's initializer that takes a throwing closure as parameter and returns .success with the value returned, or .failure with any thrown error:
> ⬇️ Suggested change
> -                              do {
>
> -                                       let feedImages = try self.feedImages(from: data)
>
> -                                       completion(.success(feedImages))
>
> -                               } catch {
>
> -                                       completion(.failure(Error.invalidData))
>
> -                               }
>
> +                               completion(Result { try self.feedImages(from: data) })
>
> In FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
> > +                          }
>
> +
>
> +                               do {
>
> +                                       let feedImages = try self.feedImages(from: data)
>
> +                                       completion(.success(feedImages))
>
> +                               } catch {
>
> +                                       completion(.failure(Error.invalidData))
>
> +                               }
>
> +
>
> +                       case .failure:
>
> +                               completion(.failure(Error.connectivity))
>
> +                       }
>
> +               }
>
> +       }
>
> +
>
> +       private func feedImages(from data: Data) throws -> [FeedImage] {
>
> We could declare this helper function as static, so we won't need a reference to self from inside the load function.
> In FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
> > @@ -19,6 +19,30 @@ public final class RemoteFeedLoader: FeedLoader {
>
>        }
>
>
>
>        public func load(completion: @escaping (FeedLoader.Result) -> Void) {
>
> -               fatalError("Must be implemented")
>
> +               client.get(from: url) { [weak self] result in
>
> +                       guard let self = self else { return }
>
> Following from my previous comment, if we remove the need to upgrade self from a weak reference, we can just check for nil instead:
> ⬇️ Suggested change
> -                      guard let self = self else { return }
>
> +                       guard self != nil else { return }
>
> In FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
> > +                          guard response.statusCode == 200 else {
>
> +                                       return completion(.failure(Error.invalidData))
>
> +                               }
>
> Consider moving this logic to a helper to keep the load function shorter and easier to read. Since this portion of code deals with mapping the result of loading, consider moving this logic to the already existing helper.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

thiagopenna commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -5,6 +5,10 @@
 import Foundation
 
 public final class RemoteFeedLoader: FeedLoader {
+	enum Constants {
```
`Constants` is an implementation detail of `RemoteFeedLoader`, so it should be marked as `private`
```suggestion
	private enum Constants {
```

thiagopenna commented on FeedAPIChallenge/Feed API/RemoteFeedLoader.swift:
```diff
@@ -20,29 +24,25 @@ public final class RemoteFeedLoader: FeedLoader {
+			return .failure(Error.invalidData)
+		}
+
+		return .success(feedResult.items.map { FeedImage($0) })
```
Suggestion: another way of using `map` in this situation is to pass the initializer as a parameter:
```suggestion
		return .success(feedResult.items.map(FeedImage.init))
```

thiagopenna reviewed (commented):
Hi @Yaro812,

Congratulations, you've passed the first challenge and got your first belt stripe ⚪️①!

✅ All tests are passing, so your implementation complies with the specs.

**[Update your status](https://iosleadessentials.slack.com/archives/CTAR0LVMM/p1629389971188200) in the community now with the following emoji to celebrate your achievement: `:white-belt-1:`**

Great job, and keep it up! Remember, we're here to help you progress and get the most out of this program.

See you in the community and in the following challenge 🙂

---

**P.S. I left some final review comments. Feel free to push changes.**

thiagopenna commented:
Well done!
